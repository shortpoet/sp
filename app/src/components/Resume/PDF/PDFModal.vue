<script>
//https://alligator.io/vuejs/vue-modal-component/

  export default {
    name: 'PDFModal',
    methods: {
      close() {
        this.$emit('close');
      },
      toPDF() {
        this.$emit('to-pdf');
      },
      toPage() {
        this.$emit('to-page');
      },
      /* istanbul ignore next */
      toCanvas() {
        this.$emit('to-canvas');
      }
    },
  };
</script>
<template>
  <transition name="modal-fade">
    <div class="custom-modal-backdrop" @click="close">
      <div class="custom-modal"
        role="dialog"
        aria-labelledby="modalTitle"
        aria-describedby="modalDescription"
      >
        <header
          class="custom-modal-header"
          id="modalTitle"
        >
          <slot name="header">
            <h4 class="heading heading-4 mr-auto">Save to PDF</h4>
            <button
              type="button"
              class="btn-ex"
              @click="close"
              aria-label="Close modal"
            >
              <font-awesome-icon icon="times" />
            </button>
          </slot>
        </header>
        <section
          class="custom-modal-body"
          id="modalDescription"
        >
          <slot name="body">
            <p>
              This does not <em>download</em> a file in the traditional sense where it is retrieved from a data server. 
              It is generated fresh by your browser based on a <b>snapshot</b> of this page as you see it. 
            </p>
            <p>
              <span class="saveme-text">Save Me!</span> - actual aspect ratio.
            </p>
            <p>
              <span class="pageme-text">Page Me!</span> - A4 format and pagination.
            </p>
          </slot>
        </section>
        <footer class="custom-modal-footer">
          <slot name="footer">
            <div class="button-drawer d-flex flex-row justify-content-around">
              <div>
                <button
                  type="button"
                  class="btn-save"
                  @click="toPDF"
                  aria-label="resume to screenshot pdf"
                >
                  Save me!
                </button>
              </div>
              <div>
                <button
                  type="button"
                  class="btn-page"
                  @click="toPage"
                  aria-label="resume to paginated pdf"
                >
                  Page me!
                </button>
              </div>
              <div>
                <a href="/">
                <button
                  type="button"
                  class="btn-close"
                  @click="close"
                  aria-label="Close modal"
                >
                  Home Page 
                </button>
                </a>
              </div>
            </div>
          </slot>
        </footer>
      </div>
    </div>
  </transition>
</template>
<style lang="scss">
</style>
