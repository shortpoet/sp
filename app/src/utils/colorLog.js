const chalk = require('chalk')
import {inspect} from 'util'

// const _log  = () => Function.prototype.bind.call(console.log, console, 'LOG_ALIAS')

export const log = (color, message) => console.log(chalk[`${color}`](`${inspect(message)}`))

export const colorLog = (message, color, background) => {
  color = color || "black";
  if (background) {
    console.log(`%c` + `${message}`, `color:` + `${color};background:${background}`)
  } else {
    console.log(`%c` + `${message}`, `color:` + `${color}`)
  }
}

// module.exports =  {
//   log,
//   colorLog
// }
