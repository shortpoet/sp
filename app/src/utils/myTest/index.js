import createWrapper from './test.utils.createWrapper'
import factory from './test.utils.factory'
import factoryShallow from './test.utils.factoryShallow'
import propFinder from './test.utils.propFinder'
import propsFinder from './test.utils.propsFinder'
import propMocker from './test.utils.propMocker'
import propsMocker from './test.utils.propsMocker'
import textMatcher from './test.utils.textMatcher'
import textMatcherFactory from './test.utils.textMatcherFactory'

export default {
  createWrapper,
  factory,
  factoryShallow,
  propFinder,
  propsFinder,
  propMocker,
  propsMocker,
  textMatcher,
  textMatcherFactory
}
