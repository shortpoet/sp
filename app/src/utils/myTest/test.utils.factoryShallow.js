import { shallowMount } from '@vue/test-utils'

export const factoryShallow = (
  component, 
  /* istanbul ignore next */
  options = {
    propsData: {},
    localVue: {},
    mocks: {},
    store: {},
    // context: {}, // only for functional components
    router: {},
    computed: {},
    stubs: {},
    slots: {},
    scopedSlots: {},
    attrs: {},
    attachToDocument: false,
    attachTo: null, //HTML Element or String
    listeners: {},
    parentComponent: {},
    provide: {}
  }) => {
    return shallowMount(component, options)
  }

module.exports = factoryShallow