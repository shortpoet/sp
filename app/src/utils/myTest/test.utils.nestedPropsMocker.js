

// // this is incomplete
// // maybe try and develop using TDD  
  
// export const nestedPropsMocker = (props) => {

//   console.log(props)

//   let propsData = {}


//   const handleArray = (props) => {
//     propsData = {}
//     props.map(prop => {
//       propsData[`${prop}`] = `sample ${prop}`
//     })
//     return {
//       propsData
//     }
//   }


//   const handleNested = (props, propsData, previousKey, nextKey) => {
//     console.log(props, propsData)

//     let firstLevel
//     if (previousKey === null && nextKey === null) {
//       firstLevel = true
//     }

//     // spread to clone so the splice works on subsequent tests
//     let values = [...props]

//     const length = parseInt(values.splice(-1, 1)[0])
//     console.log(length)
//     const iterator = [...Array(length).keys()]

//     if (firstLevel === true) {
//       previousKey = values.splice(0, 1)[0]
//       propsData[`${previousKey}`] = []
//     } else {
//       console.log('#### not first run ####')
//       console.log(propsData)
//       console.log(previousKey)
//       nextKey = values.splice(0, 1)[0]
//       console.log(nextKey)
//       propsData[`${previousKey}`][`${nextKey}`] = []
//     }


//     let keyDict

//     for (let i of iterator) {
//       keyDict = {}
//       values.map(prop => {
//         if (Array.isArray(prop)) {
//           console.log(prop)
//           console.log('#### calling handleNested from nested ####')
//           // this would only work for one level nested
//           // TODO maybe refactor for infinite recursion
//           console.log(previousKey)
//           console.log(nextKey)
//           console.log(propsData)
//           handleNested(prop, propsData, previousKey, nextKey)
//         } else {
//           console.log(prop)
//           keyDict[`${prop}`] = `sample ${prop}`
//         }
//       })
//       if (firstLevel === true) {
//         propsData[`${previousKey}`].push(keyDict)
//         firstRun = false
//       } else {
//         console.log(propsData)
//         console.log(previousKey)
//         console.log(nextKey)
//         console.log(propsData.experiences)
//         console.log(propsData.experiences.jobs)
//         propsData[`${previousKey}`][`${nextKey}`].push(keyDict)
//       }
//     }
//     return {
//       propsData
//     }
//   }

//   const isNested = props.some(p => Array.isArray(p))
//   const previousKey = null
//   const nextKey = null

//   if (isNested) {
//     console.log('#### is nested ####')
//     handleNested(props, propsData, previousKey, nextKey)
//   } else  {
//     handleArray(props)
//   }
// }
