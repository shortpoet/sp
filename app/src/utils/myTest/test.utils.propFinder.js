import {hardResume} from '@/assets/resume.js'

export const propFinder = (prop) => {
  return {
    propsData: {
      [`${prop}`]: hardResume[`${prop}`]
    }
  }
}

module.exports = propFinder