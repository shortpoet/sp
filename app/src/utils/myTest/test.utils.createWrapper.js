import { createLocalVue } from '@vue/test-utils'
import factory from './test.utils.factory'
import factoryShallow from './test.utils.factoryShallow'

// import VueRouter from 'vue-router'
import Vuex from 'vuex'
import PortalVue from 'portal-vue'

import createStore from '@/store/createStore'
// import createMockRouterConfig from '@/router/createMockRouterConfig'
// import paths from '@/router/paths.js'

import { log } from '@/utils/colorLog'
export const createWrapper  = (
  component,
  /* istanbul ignore next */
  options = {},
  // but why don't i have to ignore the empty  or even unprovided resumeStoreOtions
  // might need to figure this out but setting params to ignore for now
  resumeStoreOptions = {},
  // routerOptions = {},
  isShallow = false,
  ) => {
    const localVue = createLocalVue()
    localVue.use(Vuex)
    localVue.use(PortalVue)
    const mockStoreResume = createStore.createMocks().createStoreResumeMocks(resumeStoreOptions)
    const storeConfig = createStore.createMocks({modules: {resume: mockStoreResume}})

    // Using router is untested makes tests fail
    // and just doesn't work for now
    // localVue.use(VueRouter)
    // const routerConfig = createMockRouterConfig.createMockRouterConfig(routerOptions)
    // const router = new VueRouter(routerConfig)

    // log('green', mockStoreResume)
    // log('yellow', storeConfig.modules.resume)
    const store = new Vuex.Store(storeConfig)
    // log('red', store)
    // log('blue', store._mutations['resume/SET_RESUME_RAW'][0]())
    // console.log(options)
    return isShallow ? 
    factoryShallow(component, {
      localVue,
      store,
      // router,
      ...options
    }) : 
    factory(component, {
      localVue,
      store,
      // router,
      ...options
    })
}

// export default createWrapper
// when this was set to export default a console log(createWrapper) that was in
// test.utils.textMatcher line 19 was being executed from every module that called createWrapper
module.exports = createWrapper