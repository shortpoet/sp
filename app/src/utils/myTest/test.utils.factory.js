import { mount } from '@vue/test-utils'
import { log } from '@/utils/colorLog'
export const factory = (
  component, 
  /* istanbul ignore next */
  options = {
    propsData: {},
    localVue: {},
    mocks: {},
    store: {},
    // context: {}, // only for functional components
    // this only breaks when testing
    // probably should comment out of real func (which this is (now) hehe)
    router: {},
    computed: {},
    stubs: {},
    slots: {},
    scopedSlots: {},
    attrs: {},
    attachToDocument: false,
    attachTo: null, //HTML Element or String
    listeners: {},
    parentComponent: {},
    provide: {}
  }) => {
    // console.log(options)
    // console.log(options.mocks)
    // console.log(options.store)
    // console.log(options.localVue)
    // log('blue', options.stubs)
    // log('green', options.propsData)
    // console.log(mount)
    return mount(component, options)
  }

module.exports = factory

// export default factory