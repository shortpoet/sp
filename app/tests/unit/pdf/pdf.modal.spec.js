import * as myTest from '@/utils/myTest'
const { createWrapper } = myTest.default
import PDFModal from '@/components/Resume/PDF/PDFModal'
import { FontAwesomeIcon, FontAwesomeLayers } from '@fortawesome/vue-fontawesome'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faCircle, faFilePdf, faSave, faTimes, faRocket, faGrinTongueSquint} from '@fortawesome/free-solid-svg-icons'

describe('PDFModal.vue', () => {

  const component = PDFModal

  let wrapper
  let mountOptions = {}
  let stubs

  beforeEach(() => {
    jest.resetModules()
    jest.clearAllMocks()

    library.add(
      faCircle, faFilePdf, faSave, faTimes, faRocket
    )

    stubs = {
      FontAwesomeIcon,
      FontAwesomeLayers
    }
    mountOptions = { stubs }
  })

  describe('event emissions', () => {

    it('emits close when close button is clicked', () => {
      
      const wrapper = createWrapper(component, mountOptions)
      wrapper.find('.btn-close').trigger('click')
      expect(wrapper.emitted()).toMatchObject( {'close': [ [], [] ]} )

    })
    it('emits to-pdf and close as a follow up from toPDF in PDFButtonFloat when toPDF button is clicked', () => {
      
      const wrapper = createWrapper(component, mountOptions)
      wrapper.find('.btn-save').trigger('click')
      expect(wrapper.emitted()).toMatchObject( { 'to-pdf': [ [] ], 'close': [ [] ] } )

    })
    it('emits to-page and close as a follow up from toPage in PDFButtonFloat when toPage button is clicked', () => {
      
      const wrapper = createWrapper(component, mountOptions)
      wrapper.find('.btn-page').trigger('click')
      expect(wrapper.emitted()).toMatchObject( { 'to-page': [ [] ], 'close': [ [] ] } )

    })

    it('matches snapshot', () => {

      wrapper = createWrapper(component, mountOptions)
  
      expect(wrapper.html()).toMatchSnapshot()
  
    })

  })

})
