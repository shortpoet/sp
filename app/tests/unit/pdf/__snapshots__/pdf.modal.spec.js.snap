// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDFModal.vue event emissions matches snapshot 1`] = `
<div class="custom-modal-backdrop">
  <div role="dialog" aria-labelledby="modalTitle" aria-describedby="modalDescription" class="custom-modal">
    <header id="modalTitle" class="custom-modal-header">
      <h4 class="heading heading-4 mr-auto">Save to PDF</h4> <button type="button" aria-label="Close modal" class="btn-ex"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="times" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 352 512" class="svg-inline--fa fa-times fa-w-11">
          <path fill="currentColor" d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z" class=""></path>
        </svg></button>
    </header>
    <section id="modalDescription" class="custom-modal-body">
      <p>
        This does not <em>download</em> a file in the traditional sense where it is retrieved from a data server.
        It is generated fresh by your browser based on a <b>snapshot</b> of this page as you see it.
      </p>
      <p><span class="saveme-text">Save Me!</span> - actual aspect ratio.
      </p>
      <p><span class="pageme-text">Page Me!</span> - A4 format and pagination.
      </p>
    </section>
    <footer class="custom-modal-footer">
      <div class="button-drawer d-flex flex-row justify-content-around">
        <div><button type="button" aria-label="resume to screenshot pdf" class="btn-save">
            Save me!
          </button></div>
        <div><button type="button" aria-label="resume to paginated pdf" class="btn-page">
            Page me!
          </button></div>
        <div><a href="/"><button type="button" aria-label="Close modal" class="btn-close">
              Home Page
            </button></a></div>
      </div>
    </footer>
  </div>
</div>
`;
